File: lib/rate-limit.ts
// lib/rate-limit.ts
// ---------------------------------------------------------------
// Per-user_token daily caps. Uses our existing Supabase sessions
// table — no new dependencies, no new signups.
//
// Caps (tweak in DAILY_LIMITS):
// - chat:  30 user messages per 24h
// - voice: 3  calls per 24h
//
// These caps are intentionally generous for real use but cap the
// downside if your demo URL gets posted somewhere and 200 people
// hit it. Worst case per-user/day: ~$0.50 chat + ~$1.50 voice.
// ---------------------------------------------------------------
import { supabaseAdmin, type ChatMessage } from "./supabase";

export const DAILY_LIMITS = {
  chat: 30,   // user messages per 24h
  voice: 3,   // voice calls per 24h
};

type LimitResult = {
  allowed: boolean;
  used: number;
  limit: number;
  remaining: number;
};

function dayAgoISO() {
  return new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString();
}

export async function checkChatLimit(userToken: string): Promise<LimitResult> {
  // Sum user-role messages across all chat sessions in last 24h.
  const { data, error } = await supabaseAdmin
    .from("sessions")
    .select("transcript")
    .eq("user_token", userToken)
    .eq("type", "chat")
    .gte("created_at", dayAgoISO());

  if (error) throw new Error(`rate-limit query failed: ${error.message}`);

  let used = 0;
  for (const row of data ?? []) {
    const transcript = (row.transcript as ChatMessage[]) ?? [];
    used += transcript.filter((m) => m.role === "user").length;
  }

  const limit = DAILY_LIMITS.chat;
  return {
    allowed: used < limit,
    used,
    limit,
    remaining: Math.max(0, limit - used),
  };
}

export async function checkVoiceLimit(userToken: string): Promise<LimitResult> {
  const { count, error } = await supabaseAdmin
    .from("sessions")
    .select("id", { count: "exact", head: true })
    .eq("user_token", userToken)
    .eq("type", "voice")
    .gte("created_at", dayAgoISO());

  if (error) throw new Error(`rate-limit query failed: ${error.message}`);

  const used = count ?? 0;
  const limit = DAILY_LIMITS.voice;
  return {
    allowed: used < limit,
    used,
    limit,
    remaining: Math.max(0, limit - used),
  };
}

File: backend/lib/rate_limit.py
"""
Per-user_token daily caps using the existing `sessions` table.

Mirrors lib/rate-limit.ts from the original Next.js version exactly,
so behavior is identical across stacks.

Caps:
  - chat:  30 user messages per 24h
  - voice: 3  voice calls per 24h
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import TypedDict

from .supabase_client import supabase_admin

DAILY_LIMITS = {
    "chat": 30,   # user messages per 24h
    "voice": 3,   # voice calls per 24h
}


class LimitResult(TypedDict):
    allowed: bool
    used: int
    limit: int
    remaining: int


def _day_ago_iso() -> str:
    return (datetime.now(timezone.utc) - timedelta(days=1)).isoformat()


def check_chat_limit(user_token: str) -> LimitResult:
    """Sum user-role messages across all chat sessions in last 24h."""
    res = (
        supabase_admin.table("sessions")
        .select("transcript")
        .eq("user_token", user_token)
        .eq("type", "chat")
        .gte("created_at", _day_ago_iso())
        .execute()
    )

    used = 0
    for row in (res.data or []):
        transcript = row.get("transcript") or []
        used += sum(1 for m in transcript if m.get("role") == "user")

    limit = DAILY_LIMITS["chat"]
    return LimitResult(
        allowed=used < limit,
        used=used,
        limit=limit,
        remaining=max(0, limit - used),
    )


def check_voice_limit(user_token: str) -> LimitResult:
    """Count voice sessions in last 24h."""
    res = (
        supabase_admin.table("sessions")
        .select("id", count="exact")
        .eq("user_token", user_token)
        .eq("type", "voice")
        .gte("created_at", _day_ago_iso())
        .execute()
    )

    used = res.count or 0
    limit = DAILY_LIMITS["voice"]
    return LimitResult(
        allowed=used < limit,
        used=used,
        limit=limit,
        remaining=max(0, limit - used),
    )
